homeassistant/components/sms/gateway.py

```diff
@@ -43,30 +43,34 @@ def sms_callback(self, state_machine, callback_type, callback_data):
         _LOGGER.debug(
             "Received incoming event type:%s,data:%s", callback_type, callback_data
         )
-        entries = self.get_and_delete_all_sms(state_machine)
-        _LOGGER.debug("SMS entries:%s", entries)
-        data = list()
-
-        for entry in entries:
-            decoded_entry = gammu.DecodeSMS(entry)
-            message = entry[0]
-            _LOGGER.debug("Processing sms:%s,decoded:%s", message, decoded_entry)
-            if decoded_entry is None:
-                text = message["Text"]
-            else:
-                text = ""
-                for inner_entry in decoded_entry["Entries"]:
-                    if inner_entry["Buffer"] is not None:
-                        text = text + inner_entry["Buffer"]
-
-            event_data = dict(
-                phone=message["Number"], date=str(message["DateTime"]), message=text
-            )
-
-            _LOGGER.debug("Append event data:%s", event_data)
-            data.append(event_data)
-
-        self._hass.add_job(self._notify_incoming_sms, data)
+        try:
+            entries = self.get_and_delete_all_sms(state_machine)
+            _LOGGER.debug("SMS entries:%s", entries)
+            data = list()
+
+            for entry in entries:
+                decoded_entry = gammu.DecodeSMS(entry)
+                message = entry[0]
+                _LOGGER.debug("Processing sms:%s,decoded:%s", message, decoded_entry)
+                if decoded_entry is None:
+                    text = message["Text"]
+                else:
+                    text = ""
+                    for inner_entry in decoded_entry["Entries"]:
+                        if inner_entry["Buffer"] is not None:
+                            text = text + inner_entry["Buffer"]
+
+                event_data = dict(
+                    phone=message["Number"], date=str(message["DateTime"]), message=text
+                )
+
+                _LOGGER.debug("Append event data:%s", event_data)
+                data.append(event_data)
+
+            self._hass.add_job(self._notify_incoming_sms, data)
+
+        except Exception as ex:  # pylint: disable=broad-except
+            _LOGGER.error("Error during SMS callback:%s", ex)
```

Comment:
An integration is not allow to use `broad-except` and please use `try` only around the code which you handle the exact error

Reply:
There are several reason that this needs broad-except
1. This is callback that runs inside a gammu specific worker thread, The gammu worker thread does not protect the call to ReadDevice() and even if did, there are no good ways to report this to the log making diagnosing this almost impossible. 
Please note that this code does not run in the main hass thread and it deals with lowlevel I/O, it can fail with almost any IO call to the modem for any reason.
2. We still need a broad-except when deleting sms because the exception type is specific to lib-gammy version 1.41, If we use this exception type and then downgrade gammu we would be breaking the SMS integration with a unknown type. 
3. When an IO error occurs, the entire integration no longer works, SMS signal sensor goes down, and the service to send message will no longer send messages. :(

Reply:
In the meantime, I have created another [PR in gammu](https://github.com/gammu/python-gammu/pull/40) that will address this at the gammu level and not in the SMS platform.

Reply:
I think I have resolved all your concerns, please review.